server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -1525,19 +1529,40 @@ private CanMatchResponse canMatch(ShardSearchRequest request, boolean checkRefre
                 final boolean aliasFilterCanMatch = request.getAliasFilter().getQueryBuilder() instanceof MatchNoneQueryBuilder == false;
                 FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
                 MinAndMax<?> minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
-                final boolean canMatch;
+                boolean canMatch;
                 if (canRewriteToMatchNone(request.source())) {
                     QueryBuilder queryBuilder = request.source().query();
                     canMatch = aliasFilterCanMatch && queryBuilder instanceof MatchNoneQueryBuilder == false;
+                    canMatch = aliasFilterCanMatch && queryBuilder instanceof MatchNoneQueryBuilder == false;
                 } else {
                     // null query means match_all
                     canMatch = aliasFilterCanMatch;
                 }
+                canMatch = canMatch && canMatchSearchAfter(searchContext, minMax);
```

Comment:
Creating a new search context seems to be quite heavy operation, I think we don't need full context here since we are interested in sorts and searchAfter only:

```suggestion
                final List<SortBuilder<?>> sorts = request.source().sorts();
                final Optional<SortAndFormats> sortOpt = SortBuilder.buildSort(sorts, context);

                canMatch = canMatch && sortOpt.map(sort -> canMatchSearchAfter(sorts, 
                    SearchAfterBuilder.buildFieldDoc(sort, request.source().searchAfter()), minMax)).orElse(false);

```

Reply:
And `canMatchSearchAfter` becomes:

```
public static boolean canMatchSearchAfter(List<SortBuilder<?>> sorts, FieldDoc searchAfter, MinAndMax<?> minMax) {
}
```

Reply:
Yeah, I was thinking in similar direction, but later though, canMatch phase usage is only going to increase and so parsing and building individual sources, lets just have SarchContext in canMatch shard level requests. 
But lets see if we need more items in future from SearchContext, we can later but it makes sense to me to go with this light weight searchAfter source build as you suggested.